button: separate linkbutton

JonasBa opened the pull request:
Separate LinkButton implementation from Button. I will followup with a PR to remove the href and to implementations from button and cleanup the rest of the flow inside the component

I'm breaking this into multiple PR due how involved this component is, the following is what I plan to do:

- [x] Split LinkButton into separate implementation https://github.com/getsentry/sentry/pull/90311
- [x] Decouple href and to types from Button https://github.com/getsentry/sentry/pull/90558
- [x] Add story and drop BaseButton export https://github.com/getsentry/sentry/pull/90565/files#diff-b6af1cc909d50285895b7b06628257c301a5e4268ef597f40179a90ef2c8f65f
- [x] Split LinkButton to separate file https://github.com/getsentry/sentry/pull/90591
- [x] Refactor types to enforce either to or href https://github.com/getsentry/sentry/pull/90667
- [x] Remove the ability to pass disabled to LinkButton and tests that rely on it https://github.com/getsentry/sentry/pull/90676
- [ ] Inline StyledLinkButton Link/a rendering logic to LinkButton (see if we can do the same for button)
- [ ] Check if `<LinkButton external/>` should just render `<ExternalLink/>`
- [ ] Write docs explaining the difference about LinkButton and Button into both components

JonasBa committed:
button: separate linkbutton

button: fix tooltipprops

button: wip

TkDodo commented on static/app/components/core/button/index.tsx:
since the file is quite big, can we move the `LinkButton` to its own file?

JonasBa replied:
Let me do that in a followup PR as it will create a huge diff. But yes, 100%

TkDodo commented on static/app/components/core/button/index.tsx:
```diff
@@ -144,196 +114,132 @@ interface ButtonPropsWithAriaLabel extends BaseButtonProps {
 
       buttonTracking();
-      onClick?.(e);
+      // @ts-expect-error at this point, we don't know if the button is a button or a link
```
if we form a union of `LinkButtonProps | ButtonProps`, we need to be able to discriminate between them. Then, we could do a runtime check against that discriminator prop and have TS narrow the type, and return a matching `onClick`.

I think that discriminator could be the `to` prop, but we’d have to make it required on `LinkButton`:

```
to: string | LocationDescriptor;
```

and also declare it on `BaseButtonProps`:

```
to?: never
```

JonasBa replied:
Yes, that is the plan. I'm just breaking the process of cleaning this up into multiple steps because the types have been terrible since the beginning, and updating this will cause a lot of issues.

TkDodo commented on static/app/components/core/button/index.tsx:
```diff
@@ -144,196 +114,132 @@ interface ButtonPropsWithAriaLabel extends BaseButtonProps {
+      props.onClick?.(e);
     },
-    [disabled, busy, onClick, buttonTracking]
+    [props, buttonTracking]
```
`props` is a new object on every render, so doing memoization on it doesn’t achieve anything. Can we please remove the `useCallback` ?

JonasBa replied:
Removed it. I will remove this entire function btw, and inline the busy/disabled checks

JonasBa committed:
fixup! button: separate linkbutton

JonasBa committed:
button: drop anchor element signature (#90558)

Drop anchor element props from button element

JonasBa committed:
Merge branch 'master' into jb/button/href

JonasBa committed:
button: reorganize button (#90565)

Reorganize button component to have the two implementations close to
each other and cleanup some props and issues.

I've added a LinkButton story file, but I'm unsure if that should be its
own implementation or just extra documentation to Button (I will
probably address this later).

JonasBa committed:
button: separate href and props (#90667)

Separate href and to LinkButton prop definitions from LinkButton

JonasBa committed:
Merge branch 'master' into jb/button/href

JonasBa committed:
fix: linkbutton import

JonasBa committed:
Merge branch 'master' into jb/button/href

JonasBa committed:
fix new error instances

JonasBa committed:
fix new error instance

JonasBa commented:
I'm going to prioritize merging this, as we keep using Buttons everywhere and it creates merge on an almost daily basis. The last 3 items in the list can be tackled separately and do not block merging

JonasBa committed:
Merge branch 'master' into jb/button/href

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes
